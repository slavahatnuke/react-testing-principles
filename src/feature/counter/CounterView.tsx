type ICounterViewProps = { counter: number, increment: () => void, decrement: () => void };

// view purity
export function CounterView({counter, increment, decrement}: ICounterViewProps) {
  return (
    <>
      <h3 data-testid="counter-header">Counter ({counter})</h3>

      <button data-testid="increment" onClick={increment}>
        +
      </button>

      <button data-testid="decrement" onClick={decrement}>
        -
      </button>
    </>
  );
}