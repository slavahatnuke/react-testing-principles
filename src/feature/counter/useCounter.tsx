import {useState} from "react";

type IUseCounter = {
  counter: number,
  increment: () => void,
  decrement: () => void
};

// pure logic
export function useCounter(value: number = 0): IUseCounter {
  const [counter, setCounter] = useState(value);

  const increment = () => {
    setCounter((prevCounter) => prevCounter + 1);
  };

  const decrement = () => {
    setCounter((prevCounter) => prevCounter - 1);
  };

  return {counter, increment, decrement};
}