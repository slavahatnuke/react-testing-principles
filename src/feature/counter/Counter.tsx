import {CounterView} from "./CounterView";
import {useCounter} from "./useCounter";

type ICounterProps = { value: number };

// Composition of logic and view
export const Counter = ({value}: ICounterProps) => {

  // counter logic
  const {counter, increment, decrement} = useCounter(value);

  // counter view
  return <CounterView counter={counter} increment={increment} decrement={decrement}/>;
};


