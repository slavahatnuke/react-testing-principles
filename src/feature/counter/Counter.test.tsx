import React from 'react';
import renderer from 'react-test-renderer';
import {fireEvent, render} from '@testing-library/react';
import {Counter} from "./Counter";
import {act, renderHook} from '@testing-library/react-hooks'
import {CounterView} from "./CounterView";
import {useCounter} from "./useCounter";

// test examples and the following links:

//  https://www.robinwieruch.de/react-testing-library/
//  https://www.npmjs.com/package/@testing-library/react
//  https://jestjs.io/docs/tutorial-react
// https://www.browserstack.com/guide/unit-testing-of-react-apps-using-jest
// https://github.com/testing-library/react-hooks-testing-library
// https://testing-library.com/docs/example-react-context/

describe(Counter.name, () => {
  test('renders no crash', () => {
    render(<Counter value={0}/>);
  });

  test('increments / component testing', () => {
    const {getByTestId} = render(<Counter value={0}/>);
    const counterHeader = getByTestId('counter-header');
    const incrementButton = getByTestId('increment');

    expect(counterHeader).toHaveTextContent('Counter (0)');

    fireEvent.click(incrementButton);

    expect(counterHeader).toHaveTextContent('Counter (1)');
  });

  test('snapshot testing', () => {
    const component1 = renderer.create(
      <Counter value={0}/>
    );
    expect(component1.toJSON()).toMatchSnapshot();

    const component2 = renderer.create(
      <Counter value={12}/>
    );
    expect(component2.toJSON()).toMatchSnapshot();
  });

  test('snapshot testing / view', () => {
    const increment1 = jest.fn();
    const decrement1 = jest.fn();

    const component1 = renderer.create(
      <CounterView counter={0} increment={increment1} decrement={decrement1}/>
    );
    expect(component1.toJSON()).toMatchSnapshot();

    const increment2 = jest.fn();
    const decrement2 = jest.fn();

    const component2 = renderer.create(
      <CounterView counter={12} increment={increment2} decrement={decrement2}/>
    );

    expect(component2.toJSON()).toMatchSnapshot();
  });

  test('useCounter hook / increment & decrement', async () => {
    const {result, waitFor} = renderHook(() => useCounter(10))

    expect(result.current.counter).toBe(10)

    act(() => {
      result.current.increment()
    })

    expect(result.current.counter).toBe(11)

    act(() => {
      result.current.decrement()
    })

    expect(result.current.counter).toBe(10)

    // in case of async behavior, it could use `waitFor`

    await waitFor(() => {
      // this function is a bit foolish and does not bubble up the error
      expect(result.current.counter).toBe(10)
    })

  })
})

