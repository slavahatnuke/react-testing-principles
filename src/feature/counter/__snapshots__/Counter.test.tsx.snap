// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter snapshot testing / view 1`] = `
Array [
  <h3
    data-testid="counter-header"
  >
    Counter (
    0
    )
  </h3>,
  <button
    data-testid="increment"
    onClick={[MockFunction]}
  >
    +
  </button>,
  <button
    data-testid="decrement"
    onClick={[MockFunction]}
  >
    -
  </button>,
]
`;

exports[`Counter snapshot testing / view 2`] = `
Array [
  <h3
    data-testid="counter-header"
  >
    Counter (
    12
    )
  </h3>,
  <button
    data-testid="increment"
    onClick={[MockFunction]}
  >
    +
  </button>,
  <button
    data-testid="decrement"
    onClick={[MockFunction]}
  >
    -
  </button>,
]
`;

exports[`Counter snapshot testing 1`] = `
Array [
  <h3
    data-testid="counter-header"
  >
    Counter (
    0
    )
  </h3>,
  <button
    data-testid="increment"
    onClick={[Function]}
  >
    +
  </button>,
  <button
    data-testid="decrement"
    onClick={[Function]}
  >
    -
  </button>,
]
`;

exports[`Counter snapshot testing 2`] = `
Array [
  <h3
    data-testid="counter-header"
  >
    Counter (
    12
    )
  </h3>,
  <button
    data-testid="increment"
    onClick={[Function]}
  >
    +
  </button>,
  <button
    data-testid="decrement"
    onClick={[Function]}
  >
    -
  </button>,
]
`;
